server/src/main/java/org/elasticsearch/search/fetch/FetchPhase.java

```diff
@@ -84,12 +84,18 @@ public void execute(SearchContext context, int[] docIdsToLoad, RankDocShardInfo
         try {
             hits = buildSearchHits(context, docIdsToLoad, profiler, rankDocs);
         } finally {
-            // Always finish profiling
-            ProfileResult profileResult = profiler.finish();
-            // Only set the shardResults if building search hits was successful
-            if (hits != null) {
-                context.fetchResult().shardResult(hits, profileResult);
-                hits.decRef();
+            try {
+                // Always finish profiling
+                ProfileResult profileResult = profiler.finish();
+                // Only set the shardResults if building search hits was successful
+                if (hits != null) {
+                    context.fetchResult().shardResult(hits, profileResult);
```

Comment:
For ease of reviewing: this is the only production callsite where we pass a pooled instance into `shardResult` so changing `shardResult(..)` and this logic is relatively obviously safe.